File: src/utils/minhash_utils.py
"""MinHash utilities for near-duplicate detection using algebraic containment."""

from datasketch import MinHash
from typing import List, Set, Tuple
import json


def tokens_to_minhash(tokens: Set[str], num_perm: int = 128) -> MinHash:
    """Create a MinHash from a set of tokens."""
    m = MinHash(num_perm=num_perm)
    for token in tokens:
        m.update(token.encode("utf-8"))
    return m


def minhash_to_json(m: MinHash) -> str:
    """Serialize MinHash to JSON string for storage."""
    return json.dumps({"hashvalues": m.hashvalues.tolist(), "num_perm": m.num_perm})


def minhash_from_json(data: str) -> MinHash:
    """Deserialize MinHash from JSON string."""
    import numpy as np
    obj = json.loads(data)
    m = MinHash(num_perm=obj["num_perm"])
    m.hashvalues = np.array(obj["hashvalues"], dtype=np.uint64)
    return m


def jaccard_from_minhash(m1: MinHash, m2: MinHash) -> float:
    """Estimate Jaccard similarity from two MinHash objects."""
    return m1.jaccard(m2)


def containment_from_jaccard(
    jaccard: float,
    size_a: int,
    size_b: int
) -> float:
    """
    Compute symmetric containment from Jaccard and set sizes.

    Jaccard = |A∩B| / (|A| + |B| - |A∩B|)
    => |A∩B| = J * (|A| + |B|) / (1 + J)

    Containment = |A∩B| / min(|A|, |B|)
    """
    if jaccard <= 0:
        return 0.0
    if jaccard >= 1:
        return 1.0

    intersection = jaccard * (size_a + size_b) / (1 + jaccard)
    return intersection / min(size_a, size_b)


def shingle_text(text: str, k: int = 5) -> Set[str]:
    """Generate k-shingles from text for MinHash."""
    words = text.lower().split()
    if len(words) < k:
        return {" ".join(words)}
    return {" ".join(words[i:i+k]) for i in range(len(words) - k + 1)}


def compute_containment(
    tokens_a: Set[str],
    tokens_b: Set[str],
    num_perm: int = 128
) -> float:
    """Direct containment computation via MinHash (no LSH index needed)."""
    if not tokens_a or not tokens_b:
        return 0.0

    m1 = tokens_to_minhash(tokens_a, num_perm)
    m2 = tokens_to_minhash(tokens_b, num_perm)

    jaccard = m1.jaccard(m2)
    return containment_from_jaccard(jaccard, len(tokens_a), len(tokens_b))


def cluster_articles_by_containment(
    articles: List[Tuple[str, Set[str]]],  # [(article_id, tokens)]
    threshold: float = 0.9,
    num_perm: int = 128
) -> List[List[str]]:
    """
    Cluster articles by pairwise containment.
    Returns list of clusters (each cluster is a list of article_ids).
    Simple O(n^2) within small buckets — no LSH index needed.
    """
    if not articles:
        return []

    # Build MinHash for each
    minhashes = {}
    for aid, tokens in articles:
        minhashes[aid] = tokens_to_minhash(tokens, num_perm)

    # Union-find clustering
    parent = {aid: aid for aid, _ in articles}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(x, y):
        px, py = find(x), find(y)
        if px != py:
            parent[px] = py

    ids = [aid for aid, _ in articles]
    n = len(ids)

    for i in range(n):
        for j in range(i + 1, n):
            aid1, aid2 = ids[i], ids[j]
            m1, m2 = minhashes[aid1], minhashes[aid2]
            jaccard = m1.jaccard(m2)
            containment = containment_from_jaccard(jaccard, len(articles[i][1]), len(articles[j][1]))
            if containment >= threshold:
                union(aid1, aid2)

    # Collect clusters
    clusters = {}
    for aid in ids:
        root = find(aid)
        clusters.setdefault(root, []).append(aid)

    return list(clusters.values())

File: pipeline-rs/src/utils/minhash_utils.rs
use crate::models::MinHashSignature;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

/// Number of permutations for MinHash (default 128)
pub const DEFAULT_NUM_PERM: usize = 128;

/// MinHash implementation for near-duplicate detection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MinHash {
    hashvalues: Vec<u64>,
    num_perm: usize,
}

impl MinHash {
    pub fn new(num_perm: usize) -> Self {
        Self {
            hashvalues: vec![u64::MAX; num_perm],
            num_perm,
        }
    }

    pub fn from_hashvalues(hashvalues: Vec<u64>, num_perm: usize) -> Self {
        Self { hashvalues, num_perm }
    }

    /// Update with a token
    pub fn update(&mut self, token: &[u8]) {
        // Use a simple but effective hash function (xxhash-like)
        // For production, consider using a proper xxhash implementation
        let hash = Self::hash_token(token);
        for i in 0..self.num_perm {
            // Use different seeds for each permutation (use wrapping_mul to avoid overflow)
            let seed = (i as u64).wrapping_mul(0x9e3779b97f4a7c15);
            let perm_hash = hash.wrapping_add(seed);
            let perm_hash = Self::mix64(perm_hash);
            if perm_hash < self.hashvalues[i] {
                self.hashvalues[i] = perm_hash;
            }
        }
    }

    /// Get Jaccard similarity with another MinHash
    pub fn jaccard(&self, other: &MinHash) -> f64 {
        if self.num_perm != other.num_perm {
            return 0.0;
        }
        let matches = self.hashvalues.iter().zip(&other.hashvalues).filter(|(a, b)| a == b).count();
        matches as f64 / self.num_perm as f64
    }

    /// Serialize to JSON string
    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(&MinHashSignature {
            hashvalues: self.hashvalues.clone(),
            num_perm: self.num_perm,
        })
    }

    /// Deserialize from JSON string
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        let sig: MinHashSignature = serde_json::from_str(json)?;
        Ok(Self::from_hashvalues(sig.hashvalues, sig.num_perm))
    }

    fn hash_token(token: &[u8]) -> u64 {
        // Simple FNV-1a hash
        let mut hash: u64 = 0xcbf29ce484222325;
        for &byte in token {
            hash ^= byte as u64;
            hash = hash.wrapping_mul(0x100000001b3);
        }
        hash
    }

    fn mix64(mut x: u64) -> u64 {
        x ^= x >> 33;
        x = x.wrapping_mul(0xff51afd7ed558ccd);
        x ^= x >> 33;
        x = x.wrapping_mul(0xc4ceb9fe1a85ec53);
        x ^= x >> 33;
        x
    }
}

/// Generate k-shingles from text for MinHash
pub fn shingle_text(text: &str, k: usize) -> HashSet<String> {
    let lower = text.to_lowercase();
    let words: Vec<&str> = lower.split_whitespace().collect();
    if words.len() < k {
        if words.is_empty() {
            return HashSet::new();
        }
        return [words.join(" ")].into_iter().collect();
    }
    words
        .windows(k)
        .map(|w| w.join(" "))
        .collect()
}

/// Compute containment from Jaccard and set sizes
pub fn containment_from_jaccard(jaccard: f64, size_a: usize, size_b: usize) -> f64 {
    if jaccard <= 0.0 {
        return 0.0;
    }
    if jaccard >= 1.0 {
        return 1.0;
    }
    let intersection = jaccard * (size_a + size_b) as f64 / (1.0 + jaccard);
    intersection / (size_a.min(size_b) as f64)
}

/// Compute containment directly via MinHash
pub fn compute_containment(
    tokens_a: &HashSet<String>,
    tokens_b: &HashSet<String>,
    num_perm: usize,
) -> f64 {
    if tokens_a.is_empty() || tokens_b.is_empty() {
        return 0.0;
    }

    let mut m1 = MinHash::new(num_perm);
    let mut m2 = MinHash::new(num_perm);

    for token in tokens_a {
        m1.update(token.as_bytes());
    }
    for token in tokens_b {
        m2.update(token.as_bytes());
    }

    let jaccard = m1.jaccard(&m2);
    containment_from_jaccard(jaccard, tokens_a.len(), tokens_b.len())
}

/// Cluster articles by pairwise containment
/// Returns list of clusters (each cluster is a list of article IDs)
pub fn cluster_articles_by_containment(
    articles: Vec<(String, HashSet<String>)>, // [(article_id, tokens)]
    threshold: f64,
    num_perm: usize,
) -> Vec<Vec<String>> {
    if articles.is_empty() {
        return Vec::new();
    }

    // Build MinHash for each
    let mut minhashes: HashMap<String, MinHash> = HashMap::new();
    for (aid, tokens) in &articles {
        let mut m = MinHash::new(num_perm);
        for token in tokens {
            m.update(token.as_bytes());
        }
        minhashes.insert(aid.clone(), m);
    }

    // Union-find clustering
    let mut parent: HashMap<String, String> = articles.iter().map(|(aid, _)| (aid.clone(), aid.clone())).collect();

    fn find(parent: &mut HashMap<String, String>, x: &str) -> String {
        let mut current = x.to_string();
        while parent.get(&current).map(|p| p.as_str()) != Some(&current) {
            let p = parent[&current].clone();
            parent.insert(current.clone(), p.clone());
            current = p;
        }
        current
    }

    fn union(parent: &mut HashMap<String, String>, x: &str, y: &str) {
        let px = find(parent, x);
        let py = find(parent, y);
        if px != py {
            parent.insert(px, py);
        }
    }

    let ids: Vec<String> = articles.iter().map(|(aid, _)| aid.clone()).collect();
    let n = ids.len();

    for i in 0..n {
        for j in (i + 1)..n {
            let aid1 = &ids[i];
            let aid2 = &ids[j];
            let m1 = &minhashes[aid1];
            let m2 = &minhashes[aid2];
            let jaccard = m1.jaccard(m2);

            // Find original tokens for size calculation
            let tokens_a = articles.iter().find(|(aid, _)| aid == aid1).map(|(_, t)| t.len()).unwrap_or(0);
            let tokens_b = articles.iter().find(|(aid, _)| aid == aid2).map(|(_, t)| t.len()).unwrap_or(0);

            let containment = containment_from_jaccard(jaccard, tokens_a, tokens_b);
            if containment >= threshold {
                union(&mut parent, aid1, aid2);
            }
        }
    }

    // Collect clusters
    let mut clusters: HashMap<String, Vec<String>> = HashMap::new();
    for aid in ids {
        let root = find(&mut parent, &aid);
        clusters.entry(root).or_default().push(aid);
    }

    clusters.into_values().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shingle_text() {
        let text = "hello world this is a test";
        let shingles = shingle_text(text, 3);
        assert_eq!(shingles.len(), 4); // 6 words - 3 + 1 = 4
    }

    #[test]
    fn test_shingle_text_short() {
        let text = "hello world";
        let shingles = shingle_text(text, 5);
        assert_eq!(shingles.len(), 1);
        assert!(shingles.contains("hello world"));
    }

    #[test]
    fn test_minhash_identical() {
        let mut m1 = MinHash::new(128);
        let mut m2 = MinHash::new(128);
        m1.update(b"test");
        m2.update(b"test");
        assert_eq!(m1.jaccard(&m2), 1.0);
    }

    #[test]
    fn test_minhash_different() {
        let mut m1 = MinHash::new(128);
        let mut m2 = MinHash::new(128);
        m1.update(b"hello");
        m2.update(b"world");
        let j = m1.jaccard(&m2);
        assert!(j < 1.0);
    }

    #[test]
    fn test_containment_from_jaccard() {
        // Identical sets
        assert_eq!(containment_from_jaccard(1.0, 10, 10), 1.0);
        // No overlap
        assert_eq!(containment_from_jaccard(0.0, 10, 10), 0.0);
        // Partial - Jaccard 0.5 with sizes 10, 20 gives containment = 1.0
        let c = containment_from_jaccard(0.5, 10, 20);
        assert!((c - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_cluster_identical() {
        let tokens_a: HashSet<String> = ["a", "b", "c"].iter().map(|s| s.to_string()).collect();
        let tokens_b: HashSet<String> = ["a", "b", "c"].iter().map(|s| s.to_string()).collect();
        let articles = vec![("1".to_string(), tokens_a), ("2".to_string(), tokens_b)];
        let clusters = cluster_articles_by_containment(articles, 0.9, 128);
        assert_eq!(clusters.len(), 1);
        assert_eq!(clusters[0].len(), 2);
    }

    #[test]
    fn test_cluster_different() {
        let tokens_a: HashSet<String> = ["a", "b", "c"].iter().map(|s| s.to_string()).collect();
        let tokens_b: HashSet<String> = ["x", "y", "z"].iter().map(|s| s.to_string()).collect();
        let articles = vec![("1".to_string(), tokens_a), ("2".to_string(), tokens_b)];
        let clusters = cluster_articles_by_containment(articles, 0.9, 128);
        assert_eq!(clusters.len(), 2);
    }
}